### PR TITLE
Custom Population Structure for Model 0.1 Demography

### PR DESCRIPTION
**Pull request by patricktnast**
##Custom Population Structure for Model 0.1 Demography

### Description

- *Category*:  data artifact
- *JIRA issue*: [MIC-4113](https://jira.ihme.washington.edu/browse/MIC-4113)
- *Research reference*:[ Closed Cohort Demography](https://vivarium-research.readthedocs.io/en/latest/models/other_models/pregnancy_demography/index.html#other-models-pregnancy-demography)

### Changes and notes
 
This model uses a custom, closed population of pregnancies. 

ASFR notes state a truncated normal distribution, but IV Iron had draws from a lognormal distribution. Ali said to leave it that way for now.

Note that the incidence rates for miscarriage, abortion, and ectopic pregnancy must not be scaled by prevalence (i.e., rescaled to have the susceptible population in the denominator. The easiest way to solve this is to use the measure `raw_incidence_rate` instead, ~but it appears this needs a small change in vivarium_inputs to validate this--I will file and link a PR for that shortly.~

Per conversation with Steve, it seems a good way to move forward here is to spoof the `vivarium_input` changes manually in `loader.py`, then take a closer look at implementing a similar change in `vivarium_inputs`. I will implement it this way and mark with a TODO to remove later on.

### Verification and Testing

Artifact can be generated
Simulations can be run from generated artifact
CI passes

### CODE REVIEW

File: src/vivarium_gates_nutrition_optimization/constants/data_keys.py [hunk 1]
```diff
@@ -29,6 +29,26 @@ def log_name(self):
...
+    )
+    INCIDENCE_RATE_ECTOPIC: str = "cause.ectopic_pregnancy.raw_incidence_rate"
+
+    @property
+    def name(self):
+        return "pregnancy"
+
+    @property
+    def log_name(self):
+        return "pregnancy"
```
Comment (stevebachmeier) [inline src/vivarium_gates_nutrition_optimization/constants/data_keys.py:46]: This is fine b/c it's just one word but we typically use 

```
@property
def log_name(self):
    return self.name.replace("_", " ")
```

File: src/vivarium_gates_nutrition_optimization/data/loader.py [hunk 2]
```diff
@@ -145,7 +150,43 @@ def _load_em_from_meid(location, meid, measure):
     return vi_utils.sort_hierarchical_data(data).droplevel("location")
 
 
-# TODO - add project-specific data functions here
+##################
+# Pregnancy Data #
+##################
+
+
+def get_pregnancy_end_rate(location: str) -> pd.DataFrame:
```
Comment (stevebachmeier) [inline src/vivarium_gates_nutrition_optimization/data/loader.py:158]: I haven't heard the term "pregnancy end rate". The docs call it just "pregnancy_incidence" - why the change?

File: src/vivarium_gates_nutrition_optimization/data/loader.py [hunk 3]
```diff
@@ -145,7 +150,43 @@ def _load_em_from_meid(location, meid, measure):
...
+    asfr = get_data(data_keys.PREGNANCY.ASFR, location)
+    sbr = get_data(data_keys.PREGNANCY.SBR, location)
+    sbr = sbr.reset_index(level="year_end", drop=True).reindex(
+        asfr.index, level="year_start", fill_value=0.0
```
Comment (stevebachmeier) [inline src/vivarium_gates_nutrition_optimization/data/loader.py:162]: Should probably confirm that filling w/ 0 is appropriate. 

Can you explain what this is doing? I'm assuming this reindexing is so the `.multiply` below works but I'm confused why you reset to 'year_end' and then reindex to asfr's `year_start`

File: src/vivarium_gates_nutrition_optimization/data/loader.py [hunk 4]
```diff
@@ -145,7 +150,43 @@ def _load_em_from_meid(location, meid, measure):
...
+def load_asfr(key: str, location: str) -> pd.DataFrame:
+    asfr = load_standard_data(key, location)
+    asfr = asfr.reset_index()
+    asfr_pivot = asfr.pivot(
+        index=[col for col in metadata.ARTIFACT_INDEX_COLUMNS if col != "location"],
+        columns="parameter",
+        values="value",
+    )
+    seed = f"{key}_{location}"
+    # Keep lognormal draws from Gates IV Iron until @alibow says otherwise
```
Comment (stevebachmeier) [inline src/vivarium_gates_nutrition_optimization/data/loader.py:181]: nit: make this a `#FIXME`

File: src/vivarium_gates_nutrition_optimization/data/loader.py [hunk 5]
```diff
@@ -50,14 +51,10 @@ def get_data(lookup_key: str, location: str) -> pd.DataFrame:
...
-        # data_keys.DIARRHEA.INCIDENCE_RATE: load_standard_data,
-        # data_keys.DIARRHEA.REMISSION_RATE: load_standard_data,
-        # data_keys.DIARRHEA.CSMR: load_standard_data,
-        # data_keys.DIARRHEA.EMR: load_standard_data,
-        # data_keys.DIARRHEA.DISABILITY_WEIGHT: load_standard_data,
-        # data_keys.DIARRHEA.RESTRICTIONS: load_metadata,
+        data_keys.PREGNANCY.ASFR: load_asfr,
+        data_keys.PREGNANCY.SBR: load_sbr,
+        data_keys.PREGNANCY.INCIDENCE_RATE_MISCARRIAGE: load_standard_data,
+        data_keys.PREGNANCY.INCIDENCE_RATE_ECTOPIC: load_standard_data,
```
Comment (stevebachmeier) [inline src/vivarium_gates_nutrition_optimization/data/loader.py:57]: As we discussed, we should either confirm w/ Rajan that using `load_standard_data` in conjunvction with adding `raw_incidence_rate` to the vivarium_inputs model validation dictionary is fine or create custom functions that do all the things.

**Review by stevebachmeier — DISMISSED** [review]
Looks good! I left a few nits and will mark request changes until you sort out how to handle the raw incidence rate thing

File: src/vivarium_gates_nutrition_optimization/data/loader.py (hunk 2 above)
Comment (patricktnast) [reply src/vivarium_gates_nutrition_optimization/data/loader.py:173]: this is admittedly a c/p from IV iron--"end" I think here is meant to highlight the fact that what we are getting is not really the "true" pregnancy incidence rate (e.g. new pregnancies), but the incidence rate of the _end_ of pregnancies. I guess "incidence" is implicit in rate. Perhaps `pregnancy_end_incidence` or `end_pregnancy_incidence` would be better?

File: src/vivarium_gates_nutrition_optimization/data/loader.py (hunk 4 above)
Comment (patricktnast) [reply src/vivarium_gates_nutrition_optimization/data/loader.py:196]: I'll just delete the comment, since now it sounds like lognormal is correct.

File: src/vivarium_gates_nutrition_optimization/data/loader.py (hunk 3 above)
Comment (patricktnast) [reply src/vivarium_gates_nutrition_optimization/data/loader.py:177]: pandas gets mad about trying to reindex based on a level if both objects are `MultiIndex`. The `reset_index` just drops the `year_end` level from `sbr`, so that there is a single unambiguous level upon which to join the dataframes. Since `year_end` is uniquely specified by `year_start` in this case, it is in some sense extraneous (and the reindexing adds it back in).

I'm not sure if this is the best way to do this, but it is done the same way several times in IV Iron 
(cf `loader.py` ln 203, 529, 590), so I just followed suit.

Comment (patricktnast) [reply src/vivarium_gates_nutrition_optimization/data/loader.py:177]: probably the fillna isn't needed though

File: src/vivarium_gates_nutrition_optimization/constants/data_keys.py [hunk 6]
```diff
@@ -29,6 +29,26 @@ def log_name(self):
 POPULATION = __Population()
 
 
+class __Pregnancy(NamedTuple):
+    ASFR: str = "covariate.age_specific_fertility_rate.estimate"
+    SBR: str = "covariate.stillbirth_to_live_birth_ratio.estimate"
+    INCIDENCE_RATE_MISCARRIAGE: str = (
```
[the remark this answers is not in the document]
Comment (patricktnast) [reply src/vivarium_gates_nutrition_optimization/constants/data_keys.py]: done

**Review by rmudambi — APPROVED** [review]
